src/middle/Mir_intf.ml

```diff
@@ -0,0 +1,200 @@
+open Core_kernel
```

Comment:
You mentioned this at the meeting but I had something totally different in mind. What is the idea behind this Mir_intf abstraction? We aren't going to have multiple MIRs supported...

Reply:
The idea here isn't to hide the MIR implementation but rather to have a `Mir` module exposed. 

The way it is at the moment means that all the MIR types are in scope as soon as you open `Middle`; since `Middle` now contains a bunch of other stuff I wanted to scope them.

Having the signature contained in `Mir_intf.mli` means I can include it in `Middle.mli` to get the nested module.

Reply:
I think if we want to partition off the stuff in Middle I would go the other way - put the e.g. pretty printer in its own module, etc. I think the most important part of the Middle (and its main API) is the MIR and that should be front and center.

Reply:
Also just one meta comment - it seems like this PR could have been broken into a few smaller ones to be easier to handle - for example, one could have just dealt with the changes to error variants and semantic check, another could have refactored the Middle, another could have done the refactorings necessary to get the different compiler ends into shape to fit into the compiler module interface you wrote. It's probably too late now for this PR (definitely don't want to make you do a ton of work to try to separate them out or redo them), but just for future PR harmony :)

Reply:
Yeah - I did a bunch of stuff to `Middle` thinking I had done enough with the frontend to get the abstraction working; turns out I hadn't. It won't be a lot of work to revert `Middle` and I agree that this PR is overreaching a little!